src/v/pandaproxy/schema_registry/validation.cc

```diff
@@ -0,0 +1,411 @@
+      , _record_key_subject_name_strategy{props.record_key_subject_name_strategy
+                                            .value_or(
+                                              config::shard_local_cfg()
+                                                .record_key_subject_name_strategy())}
+      , _record_value_schema_id_validation{props
+                                             .record_value_schema_id_validation
+                                             .value_or(
+                                               config::shard_local_cfg()
+                                                 .record_value_schema_id_validation())}
+      , _record_value_subject_name_strategy{
+          props.record_value_subject_name_strategy.value_or(
+            config::shard_local_cfg().record_value_subject_name_strategy())} {}
+
+    auto validate_field(
+      field field,
+      const model::topic& topic,
+      subject_name_strategy sns,
+      const model::record& record) -> ss::future<bool> {
+        iobuf_const_parser parser(
+          field == field::key ? record.key() : record.value());
+
+        if (parser.bytes_left() < 5) {
+            vlog(
+              plog.debug,
+              "validating: topic: {}, field: {}, not enough bytes: {}",
+              topic(),
+              to_string_view(field),
+              parser.bytes_left());
+            co_return false;
+        }
+
+        auto magic = parser.consume_type<char>();
+        if (magic != 0) {
+            vlog(
+              plog.debug,
+              "validating: topic: {}, field: {}, invalid magic: {}",
+              topic(),
+              to_string_view(field),
+              magic);
+            co_return false;
+        }
+
+        auto id = schema_id{parser.consume_be_type<int32_t>()};
+
+        // Optimistically check the cache in case just the id matches
+        // This is true for Avro with TopicNameStrategy
+        if (_api->_schema_id_cache.local().has(
+              topic, field, sns, id, std::nullopt)) {
+            vlog(
+              plog.debug,
+              "validating: topic: {}, field: {}, cache hit",
+              topic(),
+              to_string_view(field),
+              magic);
+            co_return true;
+        }
+
+        // Determine the schema type
+        std::optional<canonical_schema_definition> schema;
+        try {
+            schema.emplace(co_await _api->_store->get_schema_definition(id));
+        } catch (const exception& ex) {
+            vlog(
+              plog.debug,
+              "validating: topic: {}, field: {}, schema not found: {}",
+              topic(),
+              to_string_view(field),
+              ex.message());
+            co_return false;
+        }
+
+        std::optional<std::vector<int32_t>> proto_offsets;
+        if (schema->type() == schema_type::protobuf) {
+            auto offsets = get_proto_offsets(parser);
```

Comment:
i think parser here holds a reference to the input iobuf (also passed by reference). since this line occurs after a co_await scheduling point, it is subject to lifetime issues. taking the record as value or making a copy into a non-const iobuf parser would eliminate that concern. of course, if the caller always co_awaits then it safe too, but it's really nice to have the property that a coroutine is generally analyzable without looking at all the call sites.

Reply:
Done